core-bundle/src/Menu/BackendMenuBuilder.php

```diff
@@ -30,21 +38,235 @@ class BackendMenuBuilder
      */
     private $eventDispatcher;
 
-    public function __construct(FactoryInterface $factory, EventDispatcherInterface $eventDispatcher)
+    /**
+     * @var Security
+     */
+    private $security;
+
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+
+    /**
+     * @var LogoutUrlGenerator
+     */
+    private $logoutUrlGenerator;
+
+    /**
+     * @var PreviewUrlGenerator
+     */
+    private $previewUrlGenerator;
+
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    public function __construct(FactoryInterface $factory, EventDispatcherInterface $eventDispatcher, Security $security, LogoutUrlGenerator $logoutUrlGenerator, PreviewUrlGenerator $previewUrlGenerator, RouterInterface $router, RequestStack $requestStack, TranslatorInterface $translator)
```

Comment:
Same here, these changes are breaking BC imho. Again, I don't know why anyone would have ever used that class directly but as it is not `final` it would have been technically possible.

Reply:
OT: Imho including constructor signatures in the BC promise doesn't work well with DI/composition and is only ever a problem when also utilizing inheritance. Adding `final` everywhere imo is not yet practical. Can we maybe tag classes like this with an annotation in the future? Or the constructors with `@internal`?

Reply:
I am with @m-vo on this. Constructor arguments are not covered by the BC promise, because that would make DI unnecessarily complicated. Does Symfony guarantee BC for constructors?

Reply:
Jep, it does: https://symfony.com/doc/current/contributing/code/bc.html
Adding arguments to public methods is not BC and `public function __construct()` of course is a public method. Same goes for adding `final`.

Reply:
So you want me to make all arguments optional and throw a lot of exceptions instead?

Reply:
Throwing exceptions would be just the same, it would be a BC break. BC would mean that things still work the same they did before.
But if I look at the `BackendMenuBuilder` I ask myself why it builds the previewUrl, the logoutUrl and then dispatches the event? Building these two menu items and adding them to the menu could be done in their separate listeners just like any other menu item to? Then they are new services with their own arguments.
I mean, for the previewUrl you don't need the security service and you don't need the current request. You do for the logout listener and switch user stuff, though. So that kind of smells like you're mixing responsibilities.

Reply:
It is a menu builder. Its task is to build the menu. Not to dispatch an event and use five different listeners to add the menu items.

Reply:
Well, I think it's exactly the responsibility. The builder doesn't need all of the services you're currently injecting. The ones that would like to add the navigation items do.

Reply:
Ok, feel free to refactor this. 🤷‍♂